### PR TITLE
Sizeplay v 0.5

### PR DESCRIPTION
**Pull request by Useroth**
## About The Pull Request

Due to popular requests for such functionality to exist at least in the ghost cafe, let's get started with the first iteration of sizeplay stuff.
This PR is by no means exhaustive in terms of features nor perfect in terms of code quality, and it's intended to be improved and expanded in future.

The current features include:
- An universal set_size() proc that handles everything from the pixel offsets to the body size slowdowns. The smaller the micros are the slower they become.
- A sizegun that is currently adminspawn-only, that functions pretty much as its original counterpart. Putting it in your mouth doesn't however blow your head up.
- If you're bigger than another mob by a certain factor (configurable) you might step on them (accidentally or on purpose, depending on your combat mode status).
- If you're bigger than another mob by a certain factor (configurable) you can pick them up and put them in your pocket/backpack or carry them around in your hand.

TODO:

- [ ] Clean the code up a bit, perhaps make the size thing a component.
- [ ] Add more variety to the interaction messages.
- [ ] Fix the mob_holder up a little, make it throwable and have it handle certain edge case things better (such as stuffing it into a deep fryer)

## Changelog
:cl:
add: Sizeplay mechanicsssss.
/:cl:

### CODE REVIEW

**Comment by Swiftfeather** [comment]
gou putting people in his pocket oh no

**Comment by Vizoplez** [comment]
HAIL

**Comment by unit0016** [comment]
obligatory 'does it work for storage implants' question.

**Comment by Useroth** [comment]
> 
> 
> obligatory 'does it work for storage implants' question.

It does. But it shouldn't. 👀

**Comment by zeratino** [comment]
holy shit pocket blueshield pocket blueshield

**Comment by unit0016** [comment]
> > obligatory 'does it work for storage implants' question.
> 
> It does. But it shouldn't. 👀

finally. macro ops.

**Comment by Grunnyyy** [comment]
How will this impact Gameplay? As in Natags and Sec mainly? Maybe it would be reasonable to have larger people slower aswell, due to something like, more body mass, else we might just have people stomping on someone during a chase

**Comment by Arctaisia** [comment]
Virgo Size port? :eyes:
This is nice.
Though question, will this allow us to pick up mice like you can on Vorestation-base servers?

**Comment by Oscari-pog** [comment]
> How will this impact Gameplay? As in Natags and Sec mainly? Maybe it would be reasonable to have larger people slower aswell, due to something like, more body mass, else we might just have people stomping on someone during a chase

Bigger legs - bigger steps - faster :4head:

**Comment by StarHorder** [comment]
makes me want bluespace jars again

**Comment by Gandalf2k15** [comment]
Make it so cuffed tiny people that are picked up can't escape!

**Comment by Echopsychotic** [comment]
I look forward to this and I hope it will be great for the future.

**Comment by Cheshify** [comment]
Yeah... not a fan of this concept at all. I personally don't like cosmetic changes giving players advantages/disadvantages? Whole thing makes me uneasy.

**Comment by Vizoplez** [comment]
> Yeah... not a fan of this concept at all. I personally don't like cosmetic changes giving players advantages/disadvantages? Whole thing makes me uneasy.

Counterpoint. Small sprites make you harder to hit.

:niceone:

**Comment by Vizoplez** [comment]
> makes me want bluespace jars again

there are plans to add cumjars yes yes.

**Comment by SpaceVampire** [comment]
The only part of this I personally want is the part that makes smaller people slower, because trying to click on them in combat at the moment is cancer. I think the rest should just be cosmetic without extra features added. Stepping on people, and especially putting them in pockets/backpacks seems like its prime for a lot of issues.

File: modular_skyrat/modules/sizeplay/code/mob_resize.dm [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+/mob/living/proc/set_size(new_size)
+	new_size = clamp(new_size, CONFIG_GET(number/body_size_max), CONFIG_GET(number/body_size_min))
+	if (new_size != body_size_multiplier)
+		resize = new_size / body_size_multiplier
+		body_size_multiplier = new_size
+		if(CONFIG_GET(number/body_size_slowdown_start) > 0)
+			//Basically if you take slowdown_start as SS, and slowdown_factor as SF, the thing below is (SS-body_size)/SS * SF
+			//And the result's lower boundary is capped to 0, so that bigger sizes don't provide a movespeed buff
+			var/slowdown = max(0, (CONFIG_GET(number/body_size_slowdown_start) - body_size_multiplier)/CONFIG_GET(number/body_size_slowdown_start) * CONFIG_GET(number/body_size_slowdown_factor))
+			add_or_update_variable_movespeed_modifier(/datum/movespeed_modifier/bodysize, multiplicative_slowdown = slowdown)
```
Comment (KasparoVy) [inline modular_skyrat/modules/sizeplay/code/mob_resize.dm:18]: Currently no indication of movespeed changes in the size selector at character creation, no surprises please :(

**Comment by Sigmar-alkahest** [comment]
this still being worked on or?

**Comment by Funce** [comment]
Yes this is still being worked on.

**Comment by NullDagaf** [comment]
> Yes this is still being worked on.

As in being bugfixed, or adding a new feature in?
Not gonna lie. When/if the git bot adds the stale label TWICE, that's probably a good time to either close or merge.

**Comment by ZephyrTFA** [comment]
Oops all stale. @Funce

**Comment by Gandalf2k15** [comment]
reopen once ready

**Pull request closed without merging**